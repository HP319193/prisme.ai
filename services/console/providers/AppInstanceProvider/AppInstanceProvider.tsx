import { Loading } from '@prisme.ai/design-system';
import {
  createContext,
  ReactNode,
  useState,
  useCallback,
  useEffect,
  useRef,
} from 'react';
import api from '../../utils/api';
import { useContext } from '../../utils/useContext';

export interface AppInstanceContext {
  appInstance: Prismeai.AppInstance;
  loading: boolean;
  fetchAppInstance: () => void;
  saveAppInstance: (
    app: Prismeai.AppInstance
  ) => Promise<Prismeai.AppInstance | null>;
  saving: boolean;
  uninstallApp: () => Promise<Prismeai.AppInstance | null>;
}

interface AppInstanceProviderProps {
  id: string;
  workspaceId: string;
  children: ReactNode;
}

export const appInstanceContext = createContext<AppInstanceContext | undefined>(
  undefined
);

export const useAppInstance = () =>
  useContext<AppInstanceContext>(appInstanceContext);

export const AppInstanceProvider = ({
  id,
  workspaceId,
  children,
}: AppInstanceProviderProps) => {
  const [appInstance, setAppInstance] = useState<
    AppInstanceContext['appInstance']
  >();
  const [loading, setLoading] = useState<AppInstanceContext['loading']>(true);
  const [saving, setSaving] = useState<AppInstanceContext['saving']>(false);

  const fetchAppInstance: AppInstanceContext['fetchAppInstance'] = useCallback(async () => {
    const appInstance = await api.getAppInstance(workspaceId, id);
    if (appInstance) {
      setAppInstance(appInstance);
    }
    return appInstance;
  }, [id, workspaceId]);

  const saveAppInstance: AppInstanceContext['saveAppInstance'] = useCallback(
    async (data) => {
      if (!workspaceId || !appInstance || !appInstance.slug) return null;
      setSaving(true);
      const newAppInstance = await api.saveAppInstance(
        workspaceId,
        appInstance.slug,
        data
      );
      setAppInstance(newAppInstance);
      setSaving(false);
      return newAppInstance;
    },
    [appInstance, workspaceId]
  );

  const uninstallApp: AppInstanceContext['uninstallApp'] = useCallback(async () => {
    if (!workspaceId || !appInstance || !appInstance.slug) return null;
    api.uninstallApp(workspaceId, appInstance?.slug);
    return appInstance;
  }, [appInstance, workspaceId]);

  const prevId = useRef<string>('');
  useEffect(() => {
    if (prevId.current === id) return;
    prevId.current = id;
    const initialFetch = async () => {
      setLoading(true);
      await fetchAppInstance();
      setLoading(false);
    };
    initialFetch();
  }, [fetchAppInstance, id]);

  if (loading) return <Loading />;
  if (!appInstance) return null;

  return (
    <appInstanceContext.Provider
      value={{
        appInstance,
        loading,
        fetchAppInstance,
        saveAppInstance,
        saving,
        uninstallApp,
      }}
    >
      {children}
    </appInstanceContext.Provider>
  );
};

export default AppInstanceProvider;
