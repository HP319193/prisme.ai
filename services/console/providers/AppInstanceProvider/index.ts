export * from './AppInstanceProvider';
export { default } from './AppInstanceProvider';
