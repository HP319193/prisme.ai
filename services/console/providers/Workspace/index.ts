export * from './WorkspaceProvider';
export { default } from './WorkspaceProvider';
