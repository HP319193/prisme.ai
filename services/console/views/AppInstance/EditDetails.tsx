import {
  CloseCircleOutlined,
  DeleteOutlined,
  SettingOutlined,
} from '@ant-design/icons';
import {
  Button,
  Popover,
  Schema,
  SchemaForm,
  Tabs,
} from '@prisme.ai/design-system';
import { PopoverProps } from '@prisme.ai/design-system/lib/Components/Popover';
import { useTranslation } from 'next-i18next';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useTracking } from '../../components/Tracking';
import useLocalizedText from '../../utils/useLocalizedText';
import ConfirmButton from '../../components/ConfirmButton';
import { SLUG_VALIDATION_REGEXP } from '../../utils/regex';

interface EditDetailsprops extends Omit<PopoverProps, 'content'> {
  value: any;
  onSave: (values: any) => Promise<void | Record<string, string>>;
  onDelete: () => void;
  context?: string;
  disabled?: boolean;
}

export const EditDetails = ({
  value,
  onSave,
  onDelete,
  context,
  disabled,
  onOpenChange,
  ...props
}: EditDetailsprops) => {
  const { t } = useTranslation('workspaces');
  const { localize } = useLocalizedText();
  const { trackEvent } = useTracking();
  const [open, setOpen] = useState(false);
  const [values, setValues] = useState(value);

  const configSchema: Schema = useMemo(
    () => ({
      type: 'object',
      properties: {
        slug: {
          type: 'string',
          title: t('apps.details.slug.label'),
          pattern: SLUG_VALIDATION_REGEXP.source,
          errors: {
            pattern: t('automations.save.error_InvalidSlugError'),
          },
        },
        disabled: {
          type: 'boolean',
          title: t('apps.details.disabled.label'),
          description: t('apps.details.disabled.description'),
        },
      },
    }),
    [t]
  );

  const initialOpenState = useRef(false);
  useEffect(() => {
    if (!initialOpenState.current) {
      initialOpenState.current = true;
      return;
    }
    trackEvent({
      name: `${open ? 'Open' : 'Close'} Details Panel`,
      action: 'click',
    });
  }, [open, trackEvent]);

  const onChange = useCallback(
    (schema: Schema) => (changedValues: any) => {
      const newValues = { ...values };
      Object.keys(schema?.properties || {}).forEach((k) => {
        newValues[k] = changedValues[k];
      });
      setValues(newValues);
    },
    [values]
  );

  const submit = useCallback(() => {
    onSave(values);
  }, [onSave, values]);

  const buttons = useMemo(
    () => [
      <div key="1" className="flex flex-1 justify-end !mt-2 mx-4">
        <Button variant="primary" type="submit" disabled={disabled}>
          {t('save', { ns: 'common' })}
        </Button>
      </div>,
    ],
    [disabled, t]
  );

  return (
    <Popover
      titleClassName="flex m-0 pb-0 pt-4 pl-4 pr-4"
      title={({ setOpen }) => (
        <div className="flex flex-1 justify-between">
          {t('apps.details.title')}
          <button
            onClick={() => {
              trackEvent({
                name: 'Close Details Panel by clicking button',
                action: 'click',
              });
              setOpen(false);
            }}
          >
            <CloseCircleOutlined />
          </button>
        </div>
      )}
      destroyTooltipOnHide
      content={({ setOpen }) => (
        <Tabs
          className="flex flex-1"
          items={[
            {
              key: 'config',
              label: t('blocks.builder.setup.label'),
              children: (
                <SchemaForm
                  schema={configSchema}
                  initialValues={values}
                  onChange={onChange(configSchema)}
                  onSubmit={submit}
                  buttons={buttons}
                />
              ),
              active: true,
            },
          ]}
          tabBarExtraContent={
            <ConfirmButton
              onConfirm={onDelete}
              confirmLabel={t('apps.delete.confirm', {
                name: localize(value.name),
              })}
            >
              <DeleteOutlined className="translate-y-[-2px]" />
              <span className="flex">{t('apps.delete.label')}</span>
            </ConfirmButton>
          }
        />
      )}
      overlayClassName="min-w-[50%]"
      onOpenChange={(v) => {
        setOpen(v);
        onOpenChange?.(v);
      }}
      {...props}
    >
      <button type="button" className="text-lg text-gray focus:outline-none">
        <SettingOutlined />
      </button>
    </Popover>
  );
};

export default EditDetails;
