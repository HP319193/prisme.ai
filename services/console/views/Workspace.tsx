import {
  Button,
  Col,
  MenuTab,
  Row,
  SidePanel,
  Tooltip,
} from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useEffect, useMemo, useState } from 'react';
import { EventsViewer } from '../components/EventsViewer';
import getLayout from '../layouts/WorkspaceLayout';
import AutomationsSidebar from './AutomationsSidebar';
import IconApps from '../icons/icon-apps.svgr';
import IconAutomations from '../icons/icon-automations.svgr';
import IconPages from '../icons/icon-pages.svgr';
import AppsSidebar from './AppsSidebar';
import PagesSidebar from './PagesSidebar';
import { DoubleLeftOutlined, DoubleRightOutlined } from '@ant-design/icons';
import ShareWorkspace from '../components/Share/ShareWorkspace';
import Storage from '../utils/Storage';
import { useWorkspace } from '../components/WorkspaceProvider';

export const Workspace = () => {
  const { t } = useTranslation('workspaces');
  const { workspace, fullSidebar, setFullSidebar, setShare } = useWorkspace();
  const [sidebar, setSidebar] = useState(
    Storage.get('__workpaceSidebar') || 'automations'
  );

  useEffect(() => {
    Storage.set('__workpaceSidebar', sidebar);
  }, [sidebar]);

  useEffect(() => {
    setShare({
      label: t('workspace.share'),
      component: ShareWorkspace,
    });
  }, [setShare, t]);

  const menu = useMemo(
    () => [
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconApps width={16} height={16} />
            </div>
            {t('apps.link')}
          </div>
        ),
        key: 'apps',
      },
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconAutomations width={16} height={16} />
            </div>
            {t('automations.link')}
          </div>
        ),
        key: 'automations',
      },
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconPages width={16} height={16} />
            </div>
            {t('pages.link')}
          </div>
        ),
        key: 'pages',
      },
    ],
    [t]
  );
  console.log('wp', workspace);

  if (!workspace) return null;

  return (
    <>
      <Row className="h-full">
        <Col span={14} className="flex h-full">
          <EventsViewer />
        </Col>
        <Col span={10} className="flex h-full">
          <SidePanel
            variant="squared"
            className={`
            absolute top-0 right-0 bottom-0 !bg-white
            ${fullSidebar ? 'w-[90vw] drop-shadow' : 'w-full'}
            transition-all ease-in duration-200`}
            Header={
              <div className="flex flex-row items-center h-[70px] justify-between border border-gray-200 border-solid !border-t-0">
                <MenuTab
                  items={menu}
                  selected={
                    (menu.find(({ key }) => sidebar === key) || menu[0]).key
                  }
                  onSelect={setSidebar}
                />
                <Tooltip
                  title={t(
                    fullSidebar
                      ? 'workspace.collapseSidebar'
                      : 'workspace.expandSidebar'
                  )}
                  placement="left"
                >
                  <Button
                    onClick={() => setFullSidebar(!fullSidebar)}
                    className="!text-sm"
                  >
                    {fullSidebar ? (
                      <DoubleRightOutlined
                        className="color-blue"
                        alt="workspace.collapseSidebar"
                      />
                    ) : (
                      <DoubleLeftOutlined
                        className="color-blue"
                        alt="workspace.expandSidebar"
                      />
                    )}
                  </Button>
                </Tooltip>
              </div>
            }
          >
            <>
              {sidebar === 'apps' && <AppsSidebar />}
              {sidebar === 'automations' && <AutomationsSidebar />}
              {sidebar === 'pages' && <PagesSidebar />}
            </>
          </SidePanel>
        </Col>
      </Row>
    </>
  );
};

Workspace.getLayout = getLayout;

export default Workspace;
