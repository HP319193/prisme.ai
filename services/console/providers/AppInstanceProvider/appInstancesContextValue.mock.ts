import { AppInstanceContext } from './AppInstanceProvider';

export const appInstanceContextValue: AppInstanceContext = {
  appInstance: {
    slug: 'app',
    appSlug: 'my-App',
    appName: 'My App',
    config: {},
  },
  fetchAppInstance: jest.fn(),
  loading: false,
  saveAppInstance: jest.fn(),
  saving: false,
  uninstallApp: jest.fn(),
  documentation: null,
};

export default appInstanceContextValue;
