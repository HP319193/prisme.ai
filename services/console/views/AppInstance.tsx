import {
  CodeOutlined,
  DeleteOutlined,
  EditOutlined,
  EyeOutlined,
} from '@ant-design/icons';
import { Button, Schema } from '@prisme.ai/design-system';
import { validateAppInstance } from '@prisme.ai/validation';
import { Modal, notification, PageHeader, Segmented, Tooltip } from 'antd';
import { useTranslation } from 'next-i18next';
import Head from 'next/head';
import { useRouter } from 'next/router';
import { useEffect, useMemo, useRef, useState, useCallback } from 'react';
import AppEditor from '../components/AppEditor';
import HorizontalSeparatedNav from '../components/HorizontalSeparatedNav';
import SourceEdit from '../components/SourceEdit/SourceEdit';
import EditDetails from '../layouts/EditDetails';
import getLayout from '../layouts/WorkspaceLayout';
import AppInstanceProvider, {
  useAppInstance,
} from '../providers/AppInstanceProvider';
import { useWorkspace } from '../providers/Workspace';
import { SLUG_VALIDATION_REGEXP } from '../utils/regex';
import useDirtyWarning from '../utils/useDirtyWarning';
import useLocalizedText from '../utils/useLocalizedText';

export const AppInstance = () => {
  const {
    appInstance,
    documentation,
    saveAppInstance,
    saving,
    uninstallApp,
  } = useAppInstance();
  const { workspace } = useWorkspace();
  const { localize } = useLocalizedText();
  const { t } = useTranslation('workspaces');
  const { push } = useRouter();
  const [, { photo }] = Object.entries(workspace.imports || {}).find(
    ([slug]) => slug === appInstance.slug
  ) || [, {}];
  const [value, setValue] = useState(appInstance);
  const [displaySource, setDisplaySource] = useState(false);
  const [mountSource, setMountSource] = useState(true);
  const [viewMode, setViewMode] = useState(0);
  const [dirty] = useDirtyWarning(appInstance, value);

  useEffect(() => {
    setValue(appInstance);
  }, [appInstance]);

  const save = useCallback(() => {
    saveAppInstance(value);
  }, [saveAppInstance, value]);
  // Need to get the latest version with the latest value associated
  const saveAfterChange = useRef(save);
  useEffect(() => {
    saveAfterChange.current = save;
  }, [save]);

  const saveDetails = useCallback(
    async ({ slug = '', disabled }: Prismeai.AppInstance) => {
      const { config, ...prevValue } = value;
      const newValue = {
        ...prevValue,
        slug,
        disabled,
      };
      setValue(newValue);
      try {
        // Force source to be reloaded
        setMountSource(false);
        await saveAppInstance(newValue);
        push(`/workspaces/${workspace.id}/apps/${slug}`);
        setTimeout(() => setMountSource(true), 10);
        notification.success({
          message: t('apps.saveSuccess'),
          placement: 'bottomRight',
        });
      } catch (e) {
        const error: any = e;
        if (error.details) {
          notification.error({
            message: t('apps.save.error', {
              context: Object.keys(error.details || {})[0],
            }),
            placement: 'bottomRight',
          });
          return error.details;
        }
        throw e;
      }
    },
    [push, saveAppInstance, t, value, workspace.id]
  );

  const onDelete = useCallback(() => {
    Modal.confirm({
      icon: <DeleteOutlined />,
      content: t('apps.uninstall', appInstance),
      onOk: () => {
        push(`/workspaces/${workspace.id}`);
        uninstallApp();
      },
      okText: t('apps.uninstallConfirm', appInstance),
    });
  }, [appInstance, push, t, uninstallApp, workspace.id]);

  const detailsFormSchema: Schema = useMemo(
    () => ({
      type: 'object',
      properties: {
        slug: {
          type: 'string',
          title: t('apps.details.slug.label'),
          pattern: SLUG_VALIDATION_REGEXP.source,
          errors: {
            pattern: t('automations.save.error_InvalidSlugError'),
          },
        },
        disabled: {
          type: 'boolean',
          title: t('apps.details.disabled.label'),
          description: t('apps.details.disabled.description'),
        },
      },
    }),
    [t]
  );

  const showSource = useCallback(() => {
    setDisplaySource(!displaySource);
  }, [displaySource]);

  const mergeSource = useCallback(
    (source: any) => ({
      ...value,
      ...source,
    }),
    [value]
  );
  const validateSource = useCallback(
    (json: any) => {
      const isValid = validateAppInstance(mergeSource(json));
      console.log(validateAppInstance.errors);
      return isValid;
    },
    [mergeSource]
  );

  const source = useMemo(() => {
    const { appSlug, config = {}, ...source } = value;
    return { ...source, config: config && config.value };
  }, [value]);
  const setSource = useCallback(
    (source: any) => {
      setValue(mergeSource(source));
    },
    [mergeSource]
  );

  return (
    <>
      <PageHeader
        className="h-[4rem] flex items-center"
        title={
          <HorizontalSeparatedNav>
            <HorizontalSeparatedNav.Separator>
              {photo && (
                // eslint-disable-next-line @next/next/no-img-element
                <img
                  src={photo}
                  className="mr-2 h-[1.6rem]"
                  alt={appInstance.slug}
                />
              )}
              <span className="pr-page-title">
                <div className="text-base font-bold max-w-[35vw]">
                  {appInstance.slug}
                </div>
              </span>
            </HorizontalSeparatedNav.Separator>
            <HorizontalSeparatedNav.Separator>
              <Tooltip
                title={t('details.title', { context: 'apps' })}
                placement="bottom"
              >
                <EditDetails
                  schema={detailsFormSchema}
                  value={value}
                  onSave={saveDetails}
                  onDelete={onDelete}
                  context="apps"
                  key={appInstance.slug}
                />
              </Tooltip>
            </HorizontalSeparatedNav.Separator>
            <HorizontalSeparatedNav.Separator>
              <Tooltip title={t('apps.source.help')} placement="bottom">
                <button
                  className="flex flex-row focus:outline-none items-center"
                  onClick={showSource}
                >
                  <span
                    className={`flex mr-2 ${
                      displaySource ? 'text-accent' : ''
                    }`}
                  >
                    <CodeOutlined width="1.2rem" height="1.2rem" />
                  </span>
                  <span className="flex">
                    {displaySource
                      ? t('apps.source.close')
                      : t('apps.source.label')}
                  </span>
                </button>
              </Tooltip>
            </HorizontalSeparatedNav.Separator>
          </HorizontalSeparatedNav>
        }
        extra={[
          displaySource ? (
            <Button
              onClick={save}
              className="!flex flex-row"
              variant="primary"
              disabled={!dirty || saving}
            >
              {t('apps.save')}
            </Button>
          ) : (
            documentation && (
              <div>
                <div className="ml-3">
                  <Segmented
                    key="nav"
                    options={[
                      {
                        label: t('apps.doc'),
                        value: 0,
                        icon: <EyeOutlined />,
                      },
                      {
                        label: t('apps.config'),
                        value: 1,
                        icon: <EditOutlined />,
                      },
                    ]}
                    onChange={(v) => setViewMode(+v)}
                    className="pr-segmented-accent"
                  />
                </div>
              </div>
            )
          ),
        ]}
      />
      <Head>
        <title>
          {t('page_title', {
            elementName: localize(appInstance.slug),
          })}
        </title>
      </Head>
      <div className="relative flex flex-1 bg-blue-200 h-full overflow-y-auto">
        {/*docPage && <IFrameLoader src={docPage} className="flex flex-1" />*/}
        {viewMode === 1 && (
          <div className="absolute top-0 bottom-0 left-0 right-0 bg-white">
            <AppEditor
              schema={appInstance.config.schema}
              block={appInstance.config.block}
              appId={appInstance.slug!}
              key={appInstance.slug}
            />
          </div>
        )}

        <SourceEdit
          value={source}
          onChange={setSource}
          onSave={save}
          visible={displaySource}
          mounted={mountSource}
          validate={validateSource}
        />
      </div>
    </>
  );
};

const AppInstanceWithProvider = () => {
  const {
    query: { appId },
  } = useRouter();
  const { workspace } = useWorkspace();

  return (
    <AppInstanceProvider id={`${appId}`} workspaceId={workspace.id}>
      <AppInstance />
    </AppInstanceProvider>
  );
};
AppInstanceWithProvider.getLayout = getLayout;
export default AppInstanceWithProvider;
