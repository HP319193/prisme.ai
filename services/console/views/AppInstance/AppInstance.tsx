import { CodeOutlined, EditOutlined, EyeOutlined } from '@ant-design/icons';
import { Button } from '@prisme.ai/design-system';
import { validateAppInstance } from '@prisme.ai/validation';
import { notification, PageHeader, Segmented, Tooltip } from 'antd';
import { useTranslation } from 'next-i18next';
import Head from 'next/head';
import { useRouter } from 'next/router';
import { useEffect, useMemo, useRef, useState, useCallback } from 'react';
import AppEditor from '../../components/AppEditor';
import HorizontalSeparatedNav from '../../components/HorizontalSeparatedNav';
import IFrameLoader from '../../components/IFrameLoader';
import SourceEdit, {
  ValidationError,
} from '../../components/SourceEdit/SourceEdit';
import { TrackingCategory, useTracking } from '../../components/Tracking';
import getLayout from '../../layouts/WorkspaceLayout';
import AppInstanceProvider, {
  useAppInstance,
} from '../../providers/AppInstanceProvider';
import { useWorkspace } from '../../providers/Workspace';
import { ApiError } from '../../utils/api';
import { generatePageUrl, replaceSilently } from '../../utils/urls';
import useDirtyWarning from '../../utils/useDirtyWarning';
import useLocalizedText from '../../utils/useLocalizedText';
import EditDetails from './EditDetails';

export const AppInstance = () => {
  const { appInstance, documentation, saveAppInstance, saving, uninstallApp } =
    useAppInstance();
  const { workspace } = useWorkspace();
  const { localize } = useLocalizedText();
  const { t } = useTranslation('workspaces');
  const { replace } = useRouter();
  const [, { photo }] = Object.entries(workspace.imports || {}).find(
    ([slug]) => slug === appInstance.slug
  ) || [, {}];
  const [value, setValue] = useState(appInstance);
  const [displaySource, setDisplaySource] = useState(false);
  const [mountSource, setMountSource] = useState(true);
  const [viewMode, setViewMode] = useState(documentation ? 0 : 1);
  const [dirty] = useDirtyWarning(appInstance, value);
  const { trackEvent } = useTracking();

  useEffect(() => {
    setValue(appInstance);
  }, [appInstance]);

  const save = useCallback(
    async (newValue = value) => {
      const prevSlug = appInstance.slug;
      try {
        const updated = await saveAppInstance(newValue);
        if (!updated) return null;
        notification.success({
          message: t('apps.save.toast'),
          placement: 'bottomRight',
        });
        const { slug } = updated;
        if (prevSlug !== slug) {
          replaceSilently(`/workspaces/${workspace.id}/apps/${slug}`);
        }
        return updated;
      } catch (e) {
        const { details, error } = e as ApiError;
        const description = (
          <ul>
            {details ? (
              details.map(({ path, message }: any, key: number) => (
                <li key={key}>
                  {t('openapi', {
                    context: message,
                    path: path.replace(/^\.body\./, ''),
                    ns: 'errors',
                  })}
                </li>
              ))
            ) : (
              <li>{t('apps.save.reason', { context: error })}</li>
            )}
          </ul>
        );
        notification.error({
          message: t('apps.save.error'),
          description,
          placement: 'bottomRight',
        });
      }
    },
    [appInstance.slug, saveAppInstance, t, value, workspace.id]
  );
  // Need to get the latest version with the latest value associated
  const saveAfterChange = useRef(save);
  useEffect(() => {
    saveAfterChange.current = save;
  }, [save]);

  const saveDetails = useCallback(
    async ({ slug = '', disabled }: Prismeai.AppInstance) => {
      trackEvent({
        name: 'Save Details',
        action: 'click',
      });
      const { config, ...prevValue } = value;
      const newValue = {
        ...prevValue,
        slug,
        disabled,
      };
      setValue(newValue);
      try {
        // Force source to be reloaded
        setMountSource(false);
        await save(newValue);
        setTimeout(() => setMountSource(true), 10);
        notification.success({
          message: t('apps.saveSuccess'),
          placement: 'bottomRight',
        });
      } catch (e) {
        const error: any = e;
        if (error.details) {
          notification.error({
            message: t('apps.save.error', {
              context: Object.keys(error.details || {})[0],
            }),
            placement: 'bottomRight',
          });
          return error.details;
        }
        throw e;
      }
    },
    [save, t, trackEvent, value]
  );

  const onDelete = useCallback(() => {
    replace(`/workspaces/${workspace.id}`);
    uninstallApp();
  }, [replace, uninstallApp, workspace.id]);

  const showSource = useCallback(() => {
    trackEvent({
      name: `${displaySource ? 'Hide' : 'Show'} source code`,
      action: 'click',
    });
    setDisplaySource(!displaySource);
  }, [displaySource, trackEvent]);

  const mergeSource = useCallback(
    (source: any) => ({
      ...value,
      ...source,
    }),
    [value]
  );
  const [validationError, setValidationError] = useState<ValidationError>();
  const validateSource = useCallback(
    (json: any) => {
      const isValid = validateAppInstance(mergeSource(json));
      const [error] = validateAppInstance.errors || [];
      setValidationError(error as ValidationError);
      return isValid;
    },
    [mergeSource]
  );

  const source = useMemo(() => {
    const { appSlug, config = {}, ...source } = value;
    return { ...source, config: config && config.value };
  }, [value]);
  const setSource = useCallback(
    (source: any) => {
      setValue(mergeSource(source));
    },
    [mergeSource]
  );

  return (
    <>
      <PageHeader
        className="h-[4rem] flex items-center"
        title={
          <HorizontalSeparatedNav>
            <HorizontalSeparatedNav.Separator>
              {photo && (
                // eslint-disable-next-line @next/next/no-img-element
                <img
                  src={photo}
                  className="mr-2 h-[1.6rem]"
                  alt={appInstance.slug}
                />
              )}
              <span className="pr-page-title">
                <div className="text-base font-bold max-w-[35vw]">
                  {appInstance.slug}
                </div>
              </span>
            </HorizontalSeparatedNav.Separator>
            <HorizontalSeparatedNav.Separator>
              <Tooltip title={t('apps.details.title')} placement="bottom">
                <EditDetails
                  value={value}
                  onSave={saveDetails}
                  onDelete={onDelete}
                  context="apps"
                  key={appInstance.slug}
                />
              </Tooltip>
            </HorizontalSeparatedNav.Separator>
            <HorizontalSeparatedNav.Separator>
              <Tooltip title={t('apps.source.help')} placement="bottom">
                <button
                  className="flex flex-row focus:outline-none items-center"
                  onClick={showSource}
                >
                  <span
                    className={`flex mr-2 ${
                      displaySource ? 'text-accent' : ''
                    }`}
                  >
                    <CodeOutlined width="1.2rem" height="1.2rem" />
                  </span>
                  <span className="flex">
                    {displaySource
                      ? t('apps.source.close')
                      : t('apps.source.label')}
                  </span>
                </button>
              </Tooltip>
            </HorizontalSeparatedNav.Separator>
          </HorizontalSeparatedNav>
        }
        extra={[
          displaySource ? (
            <Button
              key="save"
              onClick={() => save()}
              className="!flex flex-row"
              variant="primary"
              disabled={!dirty || saving}
            >
              {t('apps.save.label')}
            </Button>
          ) : (
            documentation && (
              <div key="doc">
                <div className="ml-3">
                  <Segmented
                    key="nav"
                    options={[
                      {
                        label: t('apps.doc'),
                        value: 0,
                        icon: <EyeOutlined />,
                      },
                      {
                        label: t('apps.config'),
                        value: 1,
                        icon: <EditOutlined />,
                      },
                    ]}
                    onChange={(v) => {
                      trackEvent({
                        name: `Show ${
                          v === 0 ? 'Documentation' : 'Configuration'
                        } tab`,
                        action: 'click',
                      });
                      setViewMode(+v);
                    }}
                    className="pr-segmented-accent"
                  />
                </div>
              </div>
            )
          ),
        ]}
      />
      <Head>
        <title>
          [{localize(workspace.name)}]{' '}
          {t('page_title', {
            elementName: localize(appInstance.slug),
          })}
        </title>
      </Head>
      <div className="relative flex flex-1 h-full overflow-y-auto">
        {documentation && documentation.workspaceSlug && documentation.slug && (
          <IFrameLoader
            src={generatePageUrl(
              documentation.workspaceSlug,
              documentation.slug
            )}
            className="flex flex-1"
          />
        )}
        {viewMode === 1 && (
          <div className="absolute top-0 bottom-0 left-0 right-0 bg-white">
            <AppEditor
              schema={appInstance.config.schema}
              block={appInstance.config.block}
              appId={appInstance.slug!}
              key={appInstance.slug}
            />
          </div>
        )}

        <SourceEdit
          value={source}
          onChange={setSource}
          onSave={() => save()}
          visible={displaySource}
          mounted={mountSource}
          validate={validateSource}
          error={validationError}
        />
      </div>
    </>
  );
};

const AppInstanceWithProvider = () => {
  const {
    query: { appId },
  } = useRouter();
  const { workspace, events } = useWorkspace();

  return (
    <TrackingCategory category="Apps">
      <AppInstanceProvider
        id={`${appId}`}
        workspaceId={workspace.id}
        events={events}
      >
        <AppInstance />
      </AppInstanceProvider>
    </TrackingCategory>
  );
};
AppInstanceWithProvider.getLayout = getLayout;
export default AppInstanceWithProvider;
