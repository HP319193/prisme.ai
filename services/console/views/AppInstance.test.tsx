import renderer from 'react-test-renderer';
import { appInstanceContext } from '../providers/AppInstanceProvider';
import { workspaceContext } from '../providers/Workspace';
import { AppInstance } from './AppInstance';
import workspaceContextValue from '../providers/Workspace/workspaceContextValue.mock';
import appInstanceContextValue from '../providers/AppInstanceProvider/appInstancesContextValue.mock';

jest.mock('../utils/useYaml', () => {
  const toJSON = jest.fn();
  const toYaml = jest.fn();
  const useYaml = jest.fn(() => ({
    toJSON,
    toYaml,
  }));
  return useYaml;
});

jest.mock('next/router', () => {
  const mock = {
    query: {
      automationId: 'foo',
    },
    replace: jest.fn(),
  };
  return {
    useRouter: () => mock,
  };
});

it('should render', () => {
  const root = renderer.create(
    <workspaceContext.Provider value={workspaceContextValue}>
      <appInstanceContext.Provider value={appInstanceContextValue}>
        <AppInstance />
      </appInstanceContext.Provider>
    </workspaceContext.Provider>
  );
  expect(root.toJSON()).toMatchSnapshot();
});
