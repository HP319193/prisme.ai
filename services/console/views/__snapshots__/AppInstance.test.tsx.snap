// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="relative flex flex-1 bg-blue-200 h-full overflow-y-auto"
/>
`;
